File: pset1/credit/credit.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <math.h>
#define BUFFERSIZE 100



long get_long();
int creditCard(long credit_n, int n_digits);

int main(void){
    int i;
    long c_num = -1;
    long tmp[2] = {0};
    while (c_num == -1){
        printf("Number: ");
        c_num = get_long();
    }
    long c_num_save = c_num;
    for (i = 1; c_num > 0 ;i++){
        if (i%2){
            tmp[1] += (c_num%10);
        }else{
            if ( (c_num%10)*2 >= 10){
                tmp[0] += (((c_num%10)*2)%10) + ((c_num%10)*2)/10;
            }else{
                tmp[0] += (c_num%10)*2;
            }
        }
        c_num = c_num/10;
    }

    printf("test:%ld\n",tmp[0] + tmp[1]);
    if ( !((tmp[0] + tmp[1]) % 10)){
        if(creditCard(c_num_save ,i-1) == 3){
            printf("VISA\n");    
        }
        if(creditCard(c_num_save ,i-1) == 2){
            printf("MASTERCARD\n");    
        }
        if(creditCard(c_num_save ,i-1) == 1){
            printf("AMEX\n");    
        }
        if(creditCard(c_num_save ,i-1) == -1){//Could be optimized...
            printf("INVALID\n");    
        }

    }else{
        printf("Invalid\n");
    }

}
//takes a credit card number and its length and returns 3--> Visa, 2 --> MasterCard, 1 --> AmericanExpress, otherwise -1
int creditCard(long credit_n, int n_digits){
    int first_two_digits = (int)(credit_n / (long)(pow(10, (double)(n_digits - 2) )));//(long)(10 ^ (n_digits - 2) )  ;
    
    //Visa == 3
    if (( n_digits == 13 || n_digits == 16) && ((first_two_digits)/10) == 4 ){
        return 3;
    }
    //MasterCard == 2
    if ( n_digits == 16 && (first_two_digits >= 51 && first_two_digits <= 55)){
        return 2;
    }
    // American Express == 1
    if (( n_digits == 15) &&(first_two_digits ==34 || first_two_digits == 37) ){
        return 1;
    }
    return -1;
}
//takes user input(string) from STDIN and returns its equivalent in long; if the string contains non digits it will return -1
long get_long(){
    char num[BUFFERSIZE];
    char *ptrStr = &num[0];
    long final = 0;
    fgets(ptrStr, 95, stdin);
    int stringLen = strlen(num) - 1; // -1 to get rid of \n char.
    for(int i = 0; i < stringLen; i++){
        if(!isdigit( *(ptrStr + i) )){
            return -1;
        }
    }
    final = atol(num);
    return final;
}

File: pset6/credit/credit.py
# we ask that your program’s last line of output be AMEX\n or MASTERCARD\n or VISA\n or INVALID\n

# American Express uses 15-digit numbers
# MasterCard uses 16-digit numbers
# Visa uses 13- and 16-digit numbers.

# American Express numbers start with 34 or 37;
# MasterCard numbers start with 51, 52, 53, 54, or 55
# Visa numbers start with 4.
def main():



    creditNumber = input("Number: ")
    #if not creditNumber.isdigit() or not( len(creditNumber) >=13 and len(creditNumber) <= 16 ) or not(luhnsAlgo(creditNumber)):
    if not ( creditNumber.isdigit() or ( len(creditNumber) >=13 and len(creditNumber) <= 16 ) or (luhnsAlgo(creditNumber)) ):
        """ Invalid """
        print("INVALID")
        exit()


    if int(creditNumber[0:2]) == 34 or int(creditNumber[0:2]) == 37:
       """ AMEX """
       print("AMEX")

    elif int(creditNumber[0:2]) >= 51 and int(creditNumber[0:2]) <= 55:
        """ MASTERCARD """
        print("MASTERCARD")

    elif int(creditNumber[0]) == 4:
        """ VISA """
        print("VISA")
    else:
        print("INVALID")





#Checks whether the credit card has a valid number.
#Returns False if the credit card number is invalid, True otherwise.
def luhnsAlgo( creditN ):
    strlen = len(creditN) - 1
    tmp = 0
    for i in range( strlen, -1, -1):

        if (strlen - i)%2 == 1:
            """ odd """
            oddN = int( creditN[i] ) * 2
            if oddN > 9:
                oddN = int((oddN/10)) + (oddN % 10)
            tmp += oddN

        else:
            """ even"""
            tmp += int( creditN[i] )

    if tmp%10 == 0:
        return True

    return False

if __name__ == "__main__":
    main()